test/mocha-junit-reporter-spec.js

```diff
@@ -20,82 +23,130 @@ var debug = require('debug')('mocha-junit-reporter:tests');
     }
 
-    runner.fail(new Test('Bar can narfle the garthog', 'can narfle the garthog', 2002), {
-      stack: options.invalidChar + 'expected garthog to be dead' + options.invalidChar
-    });
+    var test = new Test(name, fn);
 
-    runner.fail(new Test('Baz can behave like a flandip', 'can behave like a flandip', 30003), {
-      name: 'BazError',
-      message: 'expected baz to be masher, a hustler, an uninvited grasper of cone'
-    });
+    var duration = options.duration;
+    if (duration != null) {
+      // mock duration so we have consistent output
+      Object.defineProperty(test, 'duration', {
+        set: function() {
+          // do nothing
+        },
+        get: function() {
+          return duration;
+        }
+      });
+    }
 
-    runner.startSuite({
-      title: 'Another suite!',
-      tests: [1]
-    });
-    runner.pass(new Test('Another suite', 'works', 400004));
+    return test;
+  }
 
-    if (options && options.includePending) {
-      runner.startSuite({
-        title: 'Pending suite!',
-        tests: [1]
-      });
-      runner.pending(new Test('Pending suite', 'pending'));
+  function executeTestRunner(runner, options, callback) {
+    if (!callback) {
+      callback = options;
+      options = null;
     }
+    options = options || {};
+    options.invalidChar = options.invalidChar || '';
+    options.title = options.title || 'Foo Bar';
+
+    var rootSuite = runner.suite;
+
+    var suite1 = Suite.create(rootSuite, options.title);
+    suite1.addTest(createTest('can weez the juice', {
+      duration: 101
+    }));
+
+    suite1.addTest(createTest('can narfle the garthog', {duration: 2002}, function(done) {
+      var err = new Error(options.invalidChar + 'expected garthog to be dead' + options.invalidChar);
+      err.stack = 'this is where the stack would be';
+      done(err);
+    }));
+
+    suite1.addTest(createTest('can behave like a flandip', {duration: 30003}, function(done) {
+      var err = new Error('expected baz to be masher, a hustler, an uninvited grasper of cone');
+      err.name = 'BazError';
+      err.stack = 'stack';
+      done(err);
+    }));
+
+    var suite2 = Suite.create(rootSuite, 'Another suite!');
+    suite2.addTest(createTest('works', {duration: 400004}));
+
+    if (options.includePending) {
+      var pendingSuite = Suite.create(rootSuite, 'Pending suite!');
+      pendingSuite.addTest(createTest('pending', null, null));
+    }
+
+    runner.run(callback);
+  }
 
-    runner.end();
+  function assertXmlEquals(actual, expected) {
+    expect(actual).xml.to.be.valid();
+    expect(actual).xml.to.equal(expected);
   }
 
-  function verifyMochaFile(path, options) {
+  function verifyMochaFile(runner, path, options) {
     var now = (new Date()).toISOString();
     debug('verify', now);
     var output = fs.readFileSync(path, 'utf-8');
-    expect(output).xml.to.be.valid();
-    expect(output).xml.to.equal(mockXml(runner.stats, options));
-    fs.unlinkSync(path);
```

Comment:
Deleting test output now happens in afterEach so that it's guaranteed to occur even if test fails.